### PR TITLE
Feature/allow scaling with scheduler update

### PR DESCRIPTION
**Pull request by lftakakura**
Still need to fix some unit tests

### CODE REVIEW

File: controller/controller.go [hunk 1]
```diff
@@ -570,12 +571,9 @@ func UpdateSchedulerConfig(
 	operationManager *models.OperationManager,
 ) error {
 	configYAML.EnsureDefaultValues()
-
 	timeoutSec := config.GetInt("updateTimeoutSeconds")
-	lockTimeoutMS := config.GetInt("watcher.lockTimeoutMs")
-	lockKey := GetLockKey(config.GetString("watcher.lockKey"), configYAML.Name)
-	maxVersions := config.GetInt("schedulers.versions.toKeep")
-
+	timeoutDur := time.Duration(timeoutSec) * time.Second
```
Comment (ghostec) [inline controller/controller.go]: UpdateSchedulerConfig does so much work that is incredibly hard to follow :/ However, I'm not sure if it's worth looking into that at this point.

File: controller/utils.go [hunk 2]
```diff
@@ -845,3 +889,366 @@ func validateMetricsTrigger(configYAML *models.ConfigYAML, logger logrus.FieldLo
...
+	downScalingLockKey := GetDownScalingLockKey(config.GetString("watcher.lockKey"), schedulerName)
+	timeout, _ := time.ParseDuration("5ms")
+	lock, err := redisClient.EnterCriticalSection(
+		redisClient.Trace(ctx),
+		downScalingLockKey,
+		timeout,
+		0, 0,
+	)
+	if err != nil || lock != nil {
+
```
Comment (ghostec) [inline controller/utils.go]: nit: unnecessary blank lines

File: controller/utils.go [hunk 3]
```diff
@@ -845,3 +889,366 @@ func validateMetricsTrigger(configYAML *models.ConfigYAML, logger logrus.FieldLo
 	}
 	return nil
 }
+
+func schedulerAndConfigFromName(
```
Comment (ghostec) [inline controller/utils.go]: this and all functions below it are really new functions?

File: controller/controller.go [hunk 4]
```diff
@@ -671,202 +648,144 @@ waitForLock:
...
-		var kubePods *v1.PodList
-		err := mr.WithSegment(models.SegmentPod, func() error {
-			var err error
-			kubePods, err = clientset.CoreV1().Pods(schedulerName).List(metav1.ListOptions{
-				LabelSelector: labels.Set{}.AsSelector().String(),
-				FieldSelector: fields.Everything().String(),
-			})
-			return err
-		})
+	scheduler.RollingUpdateStatus = "in progress"
```
Comment (ghostec) [inline controller/controller.go]: RollingUpdateStatus should not be magic strings. Maybe make it an enum (in PG), and use them in code from consts.

File: controller/utils.go (hunk 3 above)
Comment (lftakakura) [reply controller/utils.go]: They already existed in controller. I just separated the code in functions and put it in utils

Comment (ghostec) [reply controller/utils.go]: got it! 💯

File: controller/controller.go (hunk 4 above)
Comment (lftakakura) [reply controller/controller.go]: I decided to leave it as free text. It doesn't has predefined statuses and it is more like a description (should be similar to deis). It has texts like `rolled back to version vXY.Z`.
Also, it should be only human readable.

Comment (andrehp) [reply controller/controller.go]: Having some consts would be better than having the strings directly in the code. This way it would at least be easier to avoid typos and have all possible statuses in a single place.

Comment (lftakakura) [reply controller/controller.go]: I agree.

File: models/service.go [hunk 5]
```diff
@@ -59,7 +60,7 @@ func NewService(
 	if len(configYaml.Ports) <= 0 && len(configYaml.Containers) > 0 {
 		for i, container := range configYaml.Containers {
 			for j, port := range container.Ports {
-				port.Name = fmt.Sprintf("%s-%d%d", port.Name, i, j)
+				port.Name = fmt.Sprintf("%s-%d%d", strings.Split(port.Name, "-")[0], i, j)
```
[the remark this answers is not in the document]
Comment (lftakakura) [reply models/service.go]: It had a bug.
tldr - as configYAML was passed as reference, it was concatenating over the concatenated string multiple times for each pod created creating things like tcp-00-00-00-00-00 for port names.
This only affected local development

File: controller/controller.go [hunk 6]
```diff
@@ -671,202 +664,155 @@ waitForLock:
...
-				rollErr := rollback(
-					l, roomManager, mr, db, redisClient.Client, clientset,
-					&oldConfig, maxSurge, 2*timeoutDur, createdPods, deletedPods,
-					scheduler, currentVersion)
-				if rollErr != nil {
-					l.WithError(rollErr).Debug("error during update roll back")
-					err = rollErr
-				}
-				return errors.New(errMsg)
+			if errored != nil {
```
[the remark this answers is not in the document]
Comment (lftakakura) [reply controller/controller.go]: It's not correct. Next commit will solve this

File: controller/constants.go [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+import "fmt"
+
+// SchedulerVersions rolling update statuses
+const (
+	timedoutStatus   = "timed out"
+	erroredStatus    = "error"
+	canceledStatus   = "canceld"
+	inProgressStatus = "in progress"
+	deployedStatus   = "deployed"
```
[the remark this answers is not in the document]
Comment (lftakakura) [reply controller/constants.go]: We could but it would be redundant as the table row already has the version that was deployed in a version column

File: controller/controller.go [hunk 8]
```diff
@@ -584,84 +582,79 @@ func UpdateSchedulerConfig(
...
+		return err
 	}
 
-	if configYAML.AutoScaling.Max > 0 && configYAML.AutoScaling.Min > configYAML.AutoScaling.Max {
-		return errors.New("invalid parameter: autoscaling max must be greater than min")
-	}
+	// Lock watchers so they don't scale up or down and the scheduler is not
+	// overwritten with older version on database
+	// during the databse write phase
+	globalLockKey := models.GetSchedulerLockKey(config.GetString("watcher.lockKey"), configYAML.Name)
```
Comment (ghostec) [inline controller/controller.go]: Maybe using better names might help on confusion. `GetSchedulerLockKey` could mean any type of key. `GetSchedulerScalingLockKey` is a bit more obvious.

File: controller/controller.go [hunk 9]
```diff
@@ -584,84 +582,79 @@ func UpdateSchedulerConfig(
...
-		select {
-		case <-timeout.C:
-			return errors.New("timeout while wating for redis lock")
-		case <-ticker.C:
-			if operationManager.WasCanceled() {
-				l.Warn("operation was canceled")
-				return nil
-			}
+	// Lock updates on scheduler during all the process
+	configLockKey := models.GetConfigLockKey(config.GetString("watcher.lockKey"), configYAML.Name)
```
Comment (ghostec) [inline controller/controller.go]: The same about naming goes here. `models` is not exclusive to schedulers. What do you think about `GetSchedulerConfigLockKey` ?.

File: controller/controller.go [hunk 10]
```diff
@@ -671,202 +664,152 @@ waitForLock:
...
+	releaseLock(
+		logger,
+		redisClient,
+		globalLock,
+		schedulerName,
+	)
+
+	if shouldRecreatePods {
+		// Lock down scaling so it doesn't interferer with rolling update surges
+		downScalingLockKey := models.GetDownScalingLockKey(config.GetString("watcher.lockKey"), configYAML.Name)
```
Comment (ghostec) [inline controller/controller.go]: `GetSchedulerDownScalingLockKey` :eyes:

**Comment by cscatolini** [comment]
LGTM

If all tests are fixed I think we can merge this.

@Henrod would you like to take a look before we merge?

**Comment by ghostec** [comment]
LGTM
